suite/tests/client-interface/drpttracer_SUDO-test.dll.c

```diff
@@ -0,0 +1,213 @@
+ * XXX: This version only tests whether the function of the tracer can output data. The
+ * tests of whether the output data is correct will come on the next version. This version
+ * only tests whether the function of the tracer can output data. The tests of whether the
+ * output data is correct will come on the next version.
+ */
+
+#include <string.h>
+#include "dr_api.h"
+#include "client_tools.h"
+#include "drmgr.h"
+#include "drpttracer.h"
+
+typedef struct _per_thread_t {
+    /* Initialize the tracer_handle before each syscall, and free it after each syscall.
+     */
+    void *tracer_handle;
+    /* The number of recorded syscalls. */
+    int recorded_syscall_num;
+} per_thread_t;
+
+static int tls_idx;
+
+static void
+event_exit(void);
+
+static void
+event_thread_init(void *drcontext);
+
+static void
+event_thread_exit(void *drcontext);
+
+static bool
+event_filter_syscall(void *drcontext, int sysnum);
+
+static bool
+event_pre_syscall(void *drcontext, int sysnum);
+
+static void
+event_post_syscall(void *drcontext, int sysnum);
+
+static void
+end_tracing_and_dump_trace(void *drcontext);
+
+DR_EXPORT void
+dr_init(client_id_t id)
+{
+    bool ok;
+    ok = drmgr_init();
+    CHECK(ok, "drmgr_init failed");
+
+    ok = drpttracer_init();
+    CHECK(ok, "drpttracer_init failed");
+
+    dr_register_exit_event(event_exit);
+
+    ok = drmgr_register_thread_init_event(event_thread_init) &&
+        drmgr_register_thread_exit_event(event_thread_exit) &&
+        drmgr_register_pre_syscall_event(event_pre_syscall) &&
+        drmgr_register_post_syscall_event(event_post_syscall);
+    CHECK(ok, "drmgr_register_*_event failed");
+
+    dr_register_filter_syscall_event(event_filter_syscall);
+
+    tls_idx = drmgr_register_tls_field();
+    CHECK(tls_idx > -1, "unable to reserve TLS field");
+}
+
+static void
+event_exit(void)
+{
+    drpttracer_exit();
+    dr_unregister_filter_syscall_event(event_filter_syscall);
+
+    bool ok = drmgr_unregister_thread_init_event(event_thread_init) &&
+        drmgr_unregister_thread_exit_event(event_thread_exit) &&
+        drmgr_unregister_pre_syscall_event(event_pre_syscall) &&
+        drmgr_unregister_post_syscall_event(event_post_syscall) &&
+        drmgr_unregister_tls_field(tls_idx);
+    CHECK(ok, "drmgr_unregister_*_event failed");
+
+    drmgr_exit();
+    dr_fprintf(STDERR, "all done\n");
+}
+
+static void
+event_thread_init(void *drcontext)
+{
+    per_thread_t *pt = (per_thread_t *)dr_thread_alloc(drcontext, sizeof(*pt));
+    memset(pt, 0, sizeof(*pt));
+    drmgr_set_tls_field(drcontext, tls_idx, (void *)pt);
+}
+
+static void
+event_thread_exit(void *drcontext)
+{
+    per_thread_t *pt = (per_thread_t *)drmgr_get_tls_field(drcontext, tls_idx);
+    /* If the thread's last syscall doesn't trigger post_syscall event, we need to end the
+     * tracing manually.
+     */
+    if (pt->tracer_handle != NULL) {
+        end_tracing_and_dump_trace(drcontext);
```

Comment:
Can we also output some warning string in this case?

Reply:
I can do it. But could we assume the same application at every system has same behavior?

Reply:
Are we seeing this happen for some syscall at thread exit which doesn't return to the thread?

Reply:
Yes. The exit_group syscall.
The Linux manual page shows that it will not return.
https://man7.org/linux/man-pages/man2/exit_group.2.html